src/scheduler/scheduler.rs

```diff
@@ -0,0 +1,318 @@
+    pub channel: (
+        Sender<CoordinatorMessage<C>>,
+        Receiver<CoordinatorMessage<C>>,
+    ),
+    startup: Mutex<Option<Box<dyn CoordinatorWork<C>>>>,
+    finalizer: Mutex<Option<Box<dyn CoordinatorWork<C>>>>,
+}
+
+unsafe impl<C: Context> Send for Scheduler<C> {}
+unsafe impl<C: Context> Sync for Scheduler<C> {}
+
+impl<C: Context> Scheduler<C> {
+    pub fn new() -> Arc<Self> {
+        let worker_monitor: Arc<(Mutex<()>, Condvar)> = Default::default();
+        Arc::new(Self {
+            unconstrained_works: WorkBucket::new(true, worker_monitor.clone()), // `default_bucket` is always activated
+            prepare_stage: WorkBucket::new(false, worker_monitor.clone()),
+            closure_stage: WorkBucket::new(false, worker_monitor.clone()),
+            release_stage: WorkBucket::new(false, worker_monitor.clone()),
+            final_stage: WorkBucket::new(false, worker_monitor.clone()),
+            coordinator_works: WorkBucket::new(true, worker_monitor.clone()),
+            worker_group: None,
+            worker_monitor,
+            context: None,
+            coordinator_worker: None,
+            channel: channel(),
+            startup: Mutex::new(None),
+            finalizer: Mutex::new(None),
+        })
+    }
+
+    #[inline]
+    pub fn num_workers(&self) -> usize {
+        self.worker_group.as_ref().unwrap().worker_count()
+    }
+
+    pub fn initialize(
+        self: &'static Arc<Self>,
+        num_workers: usize,
+        context: &'static C,
+        tls: OpaquePointer,
+    ) {
+        let mut self_mut = self.clone();
+        let self_mut = unsafe { Arc::get_mut_unchecked(&mut self_mut) };
+
+        self_mut.context = Some(context);
+        self_mut.coordinator_worker = Some(Worker::new(0, None, Arc::downgrade(&self)));
+        self_mut.worker_group = Some(WorkerGroup::new(num_workers, Arc::downgrade(&self)));
+        self.worker_group
+            .as_ref()
+            .unwrap()
+            .spawn_workers(tls, context);
+
+        self_mut.closure_stage.set_open_condition(move || {
+            self.unconstrained_works.is_drained()
+                && self.prepare_stage.is_drained()
+                && self.worker_group().all_parked()
+        });
+        self_mut.release_stage.set_open_condition(move || {
+            self.unconstrained_works.is_drained()
+                && self.prepare_stage.is_drained()
+                && self.closure_stage.is_drained()
+                && self.worker_group().all_parked()
+        });
+        self_mut.final_stage.set_open_condition(move || {
+            self.unconstrained_works.is_drained()
+                && self.prepare_stage.is_drained()
+                && self.closure_stage.is_drained()
+                && self.release_stage.is_drained()
+                && self.worker_group().all_parked()
+        });
+    }
+
+    pub fn initialize_worker(self: &Arc<Self>, tls: OpaquePointer) {
```

Comment:
Fixed. I wrap the coordinator-worker with a mutex and removed all the unsafe code. It'll unlikely hit the performance because the coordinator-worker is rarely used (3-4 times per GC).